Update README based on new template

hermancollin opened the pull request:
This PR aims to showcase what the new README template (https://github.com/neuropoly/axondeepseg/wiki/Template-for-Readme-of-new-models) would look like.

hermancollin committed:
Update README based on new template

hermancollin committed:
Update model specific infos

hermancollin committed:
Bumped ivadomed commit version

hermancollin committed:
Added model preview image

hermancollin committed:
Added model preview image in README

mariehbourget commented on README.md:
````diff
@@ -1,10 +1,48 @@
+## Segment with this model
+To segment an image using this model, use
+```
+axondeepseg -t SEM -i <IMG_PATH> -m <path_to_model_folder> -s <PIXEL_SIZE>
````
```suggestion
axondeepseg -t SEM -i <IMG_PATH> -s <PIXEL_SIZE>
```

Because this is a "default" ADS model, the -m argument is not required.

mariehbourget commented on README.md:
```diff
@@ -1,10 +1,48 @@
+---
+
+## Model overview
+![image of segmentation obtained from this model](sem_model_preview.png)
```
The images are in a different orientation than the original PNG, is that expected?

hermancollin replied:
Yes but only for cosmetic reasons because I wanted the figure to fit more or less in a square. If you think that's confusing I can change it.

mariehbourget replied:
No it's allright, I was just wondering

mariehbourget commented on README.md:
```diff
@@ -1,10 +1,48 @@
+- [ivadomed](https://ivadomed.org/) commit: e6554281d07b5afef9e68ce8b02e86b02bd68363
+- [axondeepseg](https://axondeepseg.readthedocs.io/en/latest/) commit: 805868e39eddf620c9f3d60d313cadffb1121bfb
+
+## Clone this repository
```
"Clone this repository" and "Get the data" are not necessary for segmentation because it is installed with ADS.

Maybe we could have 2 main sections:
1) Train and test model: (related to ivadomed)
2) Segment (related to ADS)

hermancollin replied:
I agree that this could baffle new users. I'll reorganize these sections.

hermancollin replied:
@mariehbourget I changed the structure to have 2 main sections like you suggested.

mariehbourget replied:
Thanks, that looks great!

mariehbourget reviewed (commented):
Thanks @hermancollin for the new README template, it looks really good!
I left some minor suggestions.

hermancollin committed:
Removed -m argument

Co-authored-by: mariehbourget <54086142+mariehbourget@users.noreply.github.com>

hermancollin committed:
Reorganized sections

hermancollin committed:
Added bids_config

hermancollin commented:
As discussed in the last meeting, I trained the model and the instructions work. The default model is slightly better than my retrained model, but the results are overall very similar.

Also, I did the training and everything on `bireli` so I can confirm everything works fine on there now.

mathieuboudreau commented:
@hermancollin is this PR/branch only meant as an example, or is this ready for review and merging? If it's only an example, please change it to draft mode.

mathieuboudreau commented:
Ahh, re-reading the action tasks from our group meeting refreshed my memory; I'm testing the instructions now.

mathieuboudreau commented on README.md:
```diff
@@ -1,10 +1,55 @@
+The `-m` option can be omitted in this case because this is a default built-in model.
+
+## Train and test (ivadomed)
+
```
Mention that you should either have ADS installed and the venv activated (ivadomed will be installed), or ivadomed installed seperately, to be able to follow the following steps.

hermancollin replied:
Added

mathieuboudreau commented on README.md:
```diff
@@ -1,10 +1,55 @@
+## Train and test (ivadomed)
+
+### Clone this repository
+In order to train this model, you will need the json configuration file located in this repo.
```
Say the name of the file

mathieuboudreau commented on README.md:
```diff
@@ -1,10 +1,55 @@
+### Get the data
+The SEM dataset used to train this model is hosted on github:
+- https://github.com/axondeepseg/data_axondeepseg_sem
+- commit d7884f9ab6cd8f06cba4885e752988a7706f0ede
```
If the commit hash is important, then in the instructions say `git checkout d7884f9ab6cd8f06cba4885e752988a7706f0ede` after `cd`-ing into the repo.

mathieuboudreau commented on README.md:
````diff
@@ -1,10 +1,55 @@
+```
+
+### Train this model
+To train the model, please first update the following fields in the json training config file:
````
Say the name of the file, and location too maybe.

mathieuboudreau commented on README.md:
```diff
@@ -1,10 +1,55 @@
+
+### Train this model
+To train the model, please first update the following fields in the json training config file:
+- `fname_split`: full path to the split_dataset.joblib file
```
In which of the two repos above is that file located?

mathieuboudreau replied:
Also, maybe give an example somewhere in the readme of what exactly you consider a full path (and an example of relative paths that should be avoided). Not all users know the difference.

hermancollin replied:
Well actually now that you mention it, it doesn't have to be a full path (can also be relative to the cwd when ivadomed is called) so I think I'll remove that part.

mathieuboudreau commented on README.md:
```diff
@@ -1,10 +1,55 @@
+To train the model, please first update the following fields in the json training config file:
+- `fname_split`: full path to the split_dataset.joblib file
+- `path_data`: full path to training data
+- `gpu_ids`: specific to your hardware
```
The order of these items to be changed should be the same order that they are found in the document, in my opinion

mathieuboudreau replied:
I see more than one instance of this key, do I need to change all of them or just one specific one?
<img width="261" alt="Screen Shot 2022-03-02 at 12 01 04 AM" src="https://user-images.githubusercontent.com/1421029/156292622-28b71c3d-0d9e-4d87-8de7-20996871f186.png">

mathieuboudreau replied:
Same for output and data

hermancollin replied:
hm yes the other fields are for object detection, which we don't even use. I added the "parent" key of these hopefully it's more clear now.

mathieuboudreau commented on README.md:
```diff
@@ -1,10 +1,55 @@
+- `path_data`: full path to training data
+- `gpu_ids`: specific to your hardware
+- `path_output`: where the model will be saved
+- `bids_config`: full path to the custom bids config located in `ivadomed/config/config_bids.json`
```
If we're using the ADS installation of ivadomed, the user may need additional instructions on how to find this file, as it will be located in the venv subdirectory somewhere.

mathieuboudreau replied:
E.g. in my case, it was in ./miniconda/envs/ads_venv/lib/python3.8/site-packages/ivadomed/config/config_bids.json

mathieuboudreau commented on README.md:
````diff
@@ -1,10 +1,55 @@
+### Test this model
+To test the model, use
+```
+ivadomed --test -c path_to_config_file.json
````
This is a comment both for here, and for @mariehbourget and IVADOMED.

I did this, the end of the output looked like this, but didn't really tell me 1) if the "test" "passed" or how to evaluate it.

```
2022-03-02 10:31:28.273 | INFO     | ivadomed.evaluation:evaluate:33 -
Run Evaluation on /home/GRAMES.POLYMTL.CA/u097381/test_trained_model/pred_masks

Evaluation:   0%|                                                                                                                                                        | 0/1 [00:00<?, ?it/s]Lossy conversion from float64 to uint8. Range [0.0, 3.0]. Convert image to uint8 prior to saving to suppress this warning.
Lossy conversion from float64 to uint8. Range [0.0, 3.0]. Convert image to uint8 prior to saving to suppress this warning.
/home/GRAMES.POLYMTL.CA/u097381/miniconda/envs/ads_venv/lib/python3.8/site-packages/ivadomed/evaluation.py:110: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df_results = df_results.append(results_pred, ignore_index=True)
Evaluation: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1/1 [00:12<00:00, 12.87s/it]
2022-03-02 10:31:41.145 | INFO     | ivadomed.evaluation:evaluate:115 -                             vol_pred_class0  vol_gt_class0  rvd_class0  avd_class0  dice_class0  ...  n_pred_class1  n_gt_class1  ltpr_class1  lfdr_class1  mse_class1
image_id                                                                                         ...                                                                  
sub-rat6_sample-data15_SEM     5.804167e-07   5.445902e-07   -0.065786    0.065786      0.92575  ...            252          233     0.987124     0.230769    0.099876

[1 rows x 26 columns]
u097381@bireli:~$ 
```

mathieuboudreau replied:
I don't see any file created in the directory I was in, or anything evident elsewhere (e.g. like a segmented image) that I could check if the model it "ok" or sufficiently good to use now.

mathieuboudreau replied:
It may be also missing one additional instruction on how to "package" the trained model, as the output directory from the training seems to contain a lot more files than the trained model we provide in this repo.

mathieuboudreau replied:
<img width="1349" alt="Screen Shot 2022-03-02 at 11 35 11 AM" src="https://user-images.githubusercontent.com/1421029/156394392-2ee84cee-69d6-4588-b380-1dee146b875a.png">

hermancollin replied:
Results are in *results_eval/evaluation_3Dmetrics.csv*. I'll add it in the instructions.

hermancollin replied:
The ivadomed test command runs evaluation metrics on the test set so its not a pass/fail thing

mathieuboudreau replied:
Ok! I guess we may just need to add a little bit of text so the user knows not to expect a pass/fail message. As it is, I didn't really know what to do after running the test.

mathieuboudreau reviewed (changes requested):
@hermancollin I completed a test of all your instructions to (re-)train a model here, and left individual comments. Overall I was able to do it, but had to guess a few steps to fill in the gaps.

hermancollin committed:
Added multiple suggestions for clarity

hermancollin commented:
> @hermancollin I completed a test of all your instructions to (re-)train a model here, and left individual comments. Overall I was able to do it, but had to guess a few steps to fill in the gaps.

Thank you for these detailed suggestions. I addressed all comments, hopefully instructions are more clear now.

I wrote these instructions based on SCT models, which are directly located in the `ivadomed` organization, so maybe indeed less details are required because users should already be familiar with ivadomed, whereas ADS users may be more confused with it.

mariehbourget commented on README.md:
```diff
@@ -1,10 +1,57 @@
+## Train and test (ivadomed)
+
+### Clone this repository
+This model was trained with `ivadomed`, which is included in the `axondeepseg` venv. Otherwise, make sure you have it installed separately. You will need the *model_seg_rat_axon-myelin_sem.json* configuration file located in this repo.
```
I have a question about this.
The config file in the model folder correspond to a specific commit in ivadomed (see above on line 10) and a specific dataset commit as well. So I'm not sure if it will necessarily always work on the latest ivadomed version installed in ADS (because microscopy integration in ivadomed and model training were done in-parallel and ivadomed has changed since the training).

I think that for re-training the model, we should advise to install ivadomed separately according to ivadomed documentation and checkout the specific commit.

hermancollin replied:
Yes, I agree that this would make the instructions more sustainable if the ivadomed version was to change in ADS. I added instructions to install ivadomed separately.

hermancollin committed:
Added specific ivadomed instructions

hermancollin committed:
Removed ADS venv mention in training instructions

mariehbourget commented on README.md:
```diff
@@ -1,10 +1,64 @@
+The `-m` option can be omitted in this case because this is a default built-in model.
+
+## Train and test (ivadomed)
+This model was trained and tested with `ivadomed` on commit **e65542**. We recommand to install ivadomed in a separate virtual environment to reproduce the original training steps. Inside your venv, make sure you are on the right commit:
```
Sorry, I think I spoke too fast earlier.
The commit on which the model was trained is [55fc2067](https://github.com/ivadomed/ivadomed/commit/55fc2067cbb9c97a711e32cf8b5a377fb6d517be) as mentioned in the original [README](https://github.com/axondeepseg/default-SEM-model/blob/main/README.md).

I'm not sure what the dependencies at line 10-11 represent? Are they the commits of ADS and ivadomed when we pushed the model in production?

hermancollin replied:
Ah, yes I forgot to mention this. I actually bumped the ivadomed commit to [e65542](https://github.com/ivadomed/ivadomed/commit/e6554281d07b5afef9e68ce8b02e86b02bd68363) which corresponds to the 2.9.3 release. It is indeed the commit of ivadomed we are using in ADS right now.

I only tested the training with this commit, and I assume @mathieuboudreau did too? I actually don't think training would work with this old commit because, from what I understood, the "micr" subfolders were not yet supported and the SEM dataset is up to date with the newest BIDS version. I am not entirely sure and maybe we should test it with the old commit to see if it works, but so far I think we need to either keep this newer ivadomed commit or change the dataset commit.

edit: hm actually the old "microscopy" pattern is still indexed in version 2.9.3 so it might work with this older version. I'll take a closer look.

hermancollin replied:
Ok so apparently training the model on [55fc20](https://github.com/ivadomed/ivadomed/commit/55fc2067cbb9c97a711e32cf8b5a377fb6d517be) does work even with the "micr" subfolders, which I don't really understand given its not in the bids config file in this commit:

```
{
            "name": "datatype",
            "pattern": "[/\\\\]+(anat|beh|dwi|eeg|fmap|func|ieeg|meg|perf|microscopy|ct)[/\\\\]+"
}
```

In that case I should change the ivadomed commit.

mariehbourget replied:
Ok I understand you point, we just are trying to do different things 😄. We just need to be on the same page on WHAT we want to document here.

So the question is do we want to either:
- document how to train the model TODAY
- document how the model WAS trained

My understanding of this README is to document exactly how the model WAS trained, this is why in the original README template, I put the original commit used for training for both ivadomed (55fc2067) and the dataset (1cddcc6).

mariehbourget replied:
> Ok so apparently training the model on [55fc20](https://github.com/ivadomed/ivadomed/commit/55fc2067cbb9c97a711e32cf8b5a377fb6d517be) does work even with the "micr" subfolders, which I don't really understand given its not in the bids config file in this commit

Yeah that is weird, even the "force-index" of the files [here](https://github.com/ivadomed/ivadomed/blob/55fc2067cbb9c97a711e32cf8b5a377fb6d517be/ivadomed/loader/bids_dataframe.py#L94) did not account for "micr" at the time.

mathieuboudreau replied:
> Ok I understand you point, we just are trying to do different things 😄. We just need to be on the same page on WHAT we want to document here.
> 
> So the question is do we want to either:
> 
> * document how to train the model TODAY
> * document how the model WAS trained
> 
> My understanding of this README is to document exactly how the model WAS trained, this is why in the original README template, I put the original commit used for training for both ivadomed (55fc2067) and the dataset (1cddcc6).

I don't think either exactly is what needed. I would prefer:

* document a demo of how **a** model **was** trained.

I think continuously updating how to train a model with the current versions would be difficult, and documenting how each model was trained may also create a bigger burden on us than needed. As long as there is an example that a user can follow that is not significantly different than how you would train a new model or re-train the model with the current version (where sure they might encounter a bug or two, but could reach out to IVADOMED or ADS for help), would be sufficient.

mathieuboudreau replied:
If we wanted or needed to "document" how each model was trained, then at that point I would say we would need to consider containerizing them (e.g. in a Docker or Singularity container), but I don't know if we **really** need this yet at our stage.

mariehbourget replied:
> I don't think either exactly is what needed. I would prefer:
> * document a demo of how **a** model **was** trained.

I kind of disagree with that. How to train **a** model is already covered in the microscopy tutorial in ivadomed doc. In that case, duplicating the information here seems just more work and prone to error/discrepancies.

> I think continuously updating how to train a model with the current versions would be difficult

I agree, keeping this up-to-date is burdensome and would not serve any purposes in my opinion.

> and documenting how each model was trained may also create a bigger burden on us than needed. As long as there is an example that a user can follow that is not significantly different than how you would train a new model or re-train the model with the current version (where sure they might encounter a bug or two, but could reach out to IVADOMED or ADS for help), would be sufficient.

I think that for reproducibility purposes, the least we can do today (without docker or singularity implementation) is to document the original commits. It is a fairly small burden to document on the person doing the training, but can save a lot of time in the future if we have questions about how the training was done (this happens to us before with the SEM model trained on Keras, and we were never able to reproduces the results).

mathieuboudreau replied:
> I kind of disagree with that. How to train **a** model is already covered in the microscopy tutorial in ivadomed doc. In that case, duplicating the information here seems just more work and prone to error/discrepancies.

Most of the information here is basically just that. So should we simply provide the link to the data and the commit hash in the readme, and point all instructions to the IVADOMED? What is the purpose of duplicating the instructions here, if as you say the information here is prone to discrepancies over time.

mathieuboudreau replied:
It seems if the purpose of documenting the README is to provide sufficient information to reproduce the training, and we want to avoid duplicating information on how to train a model from what is already in IVADOMED, then we should only provide the very minimal essential information here, correct?

mariehbourget replied:
> Most of the information here is basically just that. So should we simply provide the link to the data and the commit hash in the readme, and point all instructions to the IVADOMED?

For reproducibility, I like the readme as it is now, it tells the user how to reproduce the training from the config file that is embedded in the model and which field to adapt (path, gpu_id, etc), I would just change to the real commits used for training. Adding the link to the microscopy tutorial would be great as well.

Another approach is to go minimal like you said and just mention the ivadomed and data version the model was trained on.
Here are some other examples in ivadomed repos:
- https://github.com/ivadomed/model_seg_sctumor-edema-cavity_t2-t1_unet3d-multichannel
- https://github.com/ivadomed/model_seg_exvivo_gm-wm_t2_unet2d-multichannel-softseg
- https://github.com/ivadomed/mice_uqueensland_gm

> What is the purpose of duplicating the instructions here, if as you say the information here is prone to discrepancies over time.

Sorry, that was not clear, I just meant that we should not keep the commits/versions documented in this repo up-to-date each time we update ivadomed or ADS.

hermancollin committed:
Added model output directory

hermancollin committed:
Changed "test" for "evaluate" (more clear)

jcohenadad commented on README.md:
```diff
@@ -1,10 +1,65 @@
 AxonDeepSeg default SEM model and testing image. This model works at a resolution of 0.1 micrometer per pixel and was trained on rat spinal cord data collected via a Scanning Electron Microscope (SEM).
 
+## Dependencies
+- [ivadomed](https://ivadomed.org/) commit: e6554281d07b5afef9e68ce8b02e86b02bd68363
+- [axondeepseg](https://axondeepseg.readthedocs.io/en/latest/) commit: 805868e39eddf620c9f3d60d313cadffb1121bfb
```
no need to write it twice-- the SHA are already written later 
you can just remove this section

jcohenadad commented:
Ultimately, ivadomed should include most reproducibility information in its JSON file, and the JSON file would be uploaded alongside the model. See: https://github.com/ivadomed/ivadomed/issues/167#issuecomment-1058417377

the benefits is that with the currently proposed approach (README written by hand), errors are possible

mariehbourget commented:
> Ultimately, ivadomed should include most reproducibility information in its JSON file, and the JSON file would be uploaded alongside the model. See: [ivadomed/ivadomed#167 (comment)](https://github.com/ivadomed/ivadomed/issues/167#issuecomment-1058417377)
> 
> the benefits is that with the currently proposed approach (README written by hand), errors are possible

That's a great point, thank you @jcohenadad!
I checked the model's output and we have a `version_info.log` text file with the commits and training SHA information.

Example attached:
[version_info.zip](https://github.com/axondeepseg/default-SEM-model/files/8180740/version_info.zip)

I also found that I may have put the wrong commits in the first place in the README... I'll have to do some checks and get back on that. But yeah, that is a great case for taking the automatic version log and not copy/pasting info 😬.

hermancollin committed:
Removed dependencies section (duplicate)

hermancollin commented:
> > Ultimately, ivadomed should include most reproducibility information in its JSON file, and the JSON file would be uploaded alongside the model. See: [ivadomed/ivadomed#167 (comment)](https://github.com/ivadomed/ivadomed/issues/167#issuecomment-1058417377)
> > the benefits is that with the currently proposed approach (README written by hand), errors are possible
> 
> That's a great point, thank you @jcohenadad! I checked the model's output and we have a `version_info.log` text file with the commits and training SHA information.
> 
> Example attached: [version_info.zip](https://github.com/axondeepseg/default-SEM-model/files/8180740/version_info.zip)
> 
> I also found that I may have put the wrong commits in the first place in the README... I'll have to do some checks and get back on that. But yeah, that is a great case for taking the automatic version log and not copy/pasting info :grimacing:.

According to this version file, commits for both ivadomed and the dataset itself are different than what I wrote as well as the ones in the original readme.

Following these discussions, maybe the most robust way to go for the new readme template would be to
- add the version files in the repos
- keep the instructions minimal and add a note that the exact training commits are documented in the version files for reproducibility
- direct the user to the microscopy training tutorial for more information

mariehbourget commented:
> According to this version file, commits for both ivadomed and the dataset itself are different than what I wrote as well as the ones in the original readme.

The commits in the version file I sent before make no sense. I just realized that each time we run ivadomed (either for training, testing or segment), the version file is overwritten.
Fortunately, I have a backup of the original version file, attached.
[version_info.zip](https://github.com/axondeepseg/default-SEM-model/files/8181507/version_info.zip)

The commits correspond to the one we had in the original README.

hermancollin committed:
Added version file and removed revision SHA

jcohenadad committed:
Delete .DS_Store

hermancollin commented:
The changes discussed here and during today's meeting were added. I will wait for this PR to be approved by everyone before changing the other models readme and the template.

mariehbourget commented on README.md:
````diff
@@ -1,10 +1,47 @@
+```
+
+### Get the data
+The SEM dataset used to train this model is hosted on github [here](https://github.com/axondeepseg/data_axondeepseg_sem). The specific dataset revision hash is documented in the *version_info.log* file.
````
```suggestion
The SEM dataset used to train this model is hosted on github [here](https://github.com/axondeepseg/data_axondeepseg_sem). The specific dataset revision hash used for training is documented in the *version_info.log* file.
```

hermancollin committed:
Added MH's suggestion

Co-authored-by: mariehbourget <54086142+mariehbourget@users.noreply.github.com>

mariehbourget reviewed (approved):
LGTM, thank you @hermancollin!

hermancollin commented:
I will wait for @mathieuboudreau's final take before moving on to the other models repo. I think all the suggestions you made were adressed.

Thank you everyone for your involvement in this issue.

mathieuboudreau committed:
Capitalize GitHub

mathieuboudreau committed:
Minor fixes

mathieuboudreau reviewed (approved):
I added a few minor changes, looks good to go to me!
